workflow-runtime/src/main/java/com/squareup/workflow1/WorkflowInterceptor.kt

```diff
@@ -165,3 +264,55 @@ internal fun <P, S, O, R> WorkflowInterceptor.intercept(
     override fun toString(): String = "InterceptedWorkflow($workflow, $this@intercept)"
   }
 }
+
+@OptIn(ExperimentalWorkflowApi::class)
+private class InterceptedRenderContext<P, S, O>(
+  private val baseRenderContext: BaseRenderContext<P, S, O>,
+  private val interceptor: RenderContextInterceptor<P, S, O>
+) : BaseRenderContext<P, S, O>, Sink<WorkflowAction<P, S, O>> {
+  override val actionSink: Sink<WorkflowAction<P, S, O>> get() = this
+
+  override fun send(value: WorkflowAction<P, S, O>) {
+    interceptor.onActionSent(value) { interceptedAction ->
+      baseRenderContext.actionSink.send(interceptedAction)
+    }
+  }
+
+  override fun <ChildPropsT, ChildOutputT, ChildRenderingT> renderChild(
+    child: Workflow<ChildPropsT, ChildOutputT, ChildRenderingT>,
+    props: ChildPropsT,
+    key: String,
+    handler: (ChildOutputT) -> WorkflowAction<P, S, O>
+  ): ChildRenderingT =
+    interceptor.onRenderChild(child, props, key, handler) { c, p, k, h ->
+      // TODO better lambda arg names
+      baseRenderContext.renderChild(c, p, k, h)
+    }
+
+  override fun runningSideEffect(
+    key: String,
+    sideEffect: suspend CoroutineScope.() -> Unit
+  ) {
+    baseRenderContext.runningSideEffect(key) {
+      val outerScope = this
+
+      // Only invoke the hook when the side effect is actually starting – i.e. don't invoke
+      // it on subsequent consecutive calls with the same key.
+      interceptor.onSideEffectRunning(key) {
+        // We don't want to use the CoroutineScope receiver from the outer runningSideEffect
+        // here since the interceptor could be calling this function from a different context
+        // (e.g. wrap it with a different dispatcher or name).
+        coroutineScope {
+          val innerScope = this
+          // If the interceptor is wrapping the coroutine context, it can't change the job
+          // because that could break structured concurrency.
+          check(outerScope.coroutineContext.job === innerScope.coroutineContext.job) {
```

Comment:
This can't succeed, because `coroutineScope` introduces a new `Job`:

```
 * The provided scope inherits its [coroutineContext][CoroutineScope.coroutineContext] from the outer scope, but overrides
 * the context's [Job].
```

Still figuring out how to fix it.

Reply:
I'm stuck. Tried this:

```kotlin
    baseRenderContext.runningSideEffect(key) {
      val outerScope = this
      println("job 1: $coroutineContext, ${coroutineContext.job}")

      // Only invoke the hook when the side effect is actually starting – i.e. don't invoke
      // it on subsequent consecutive calls with the same key.
      interceptor.onSideEffectRunning(key) {
        println("job 4: $coroutineContext, ${coroutineContext.job}")
        // If the interceptor is wrapping the coroutine context, it can't change the job
        // because that could break structured concurrency.
        check(outerScope.coroutineContext.job === coroutineContext.job) {
          "Expected onSideEffectRunning not to call proceed with a different Job."
        }

        // We don't want to use the CoroutineScope receiver from the outer runningSideEffect
        // here since the interceptor could be calling this function from a different context
        // (e.g. wrap it with a different dispatcher or name).
        coroutineScope {
          sideEffect()
        }
      }
    }
```

With a `RenderContextInterceptor` that looks like this:

```kotlin
          override suspend fun onSideEffectRunning(
            key: String,
            proceed: suspend () -> Unit
          ) {
            println("job 2: $coroutineContext, ${coroutineContext.job}")
            // Creates a new Job, so sneaky.
            coroutineScope {
              println("job 3: $coroutineContext, ${coroutineContext.job}")
              proceed()
            }
          }
```

Prints this, revealing that `check()` call is just going to capture the
context that `runningSideEffect` was called in, it's blind to what the
interceptor method does behind its back. I don't see how I can detect it.

```
job 1: [CoroutineId(1), "coroutine#1":BlockingCoroutine{Active}@79efed2d, BlockingEventLoop@2928854b], "coroutine#1":BlockingCoroutine{Active}@79efed2d
job 2: [CoroutineId(1), "coroutine#1":BlockingCoroutine{Active}@79efed2d, BlockingEventLoop@2928854b], "coroutine#1":BlockingCoroutine{Active}@79efed2d
job 3: [CoroutineId(1), "coroutine#1":ScopeCoroutine{Active}@2bbf180e, BlockingEventLoop@2928854b], "coroutine#1":ScopeCoroutine{Active}@2bbf180e
job 4: [CoroutineId(1), "coroutine#1":BlockingCoroutine{Active}@79efed2d, BlockingEventLoop@2928854b], "coroutine#1":BlockingCoroutine{Active}@79efed2d
```

Reply:
Ooh, maybe I can check the job's super job? Is that a thing?

Reply:
Hot doggy!

```
          check(innerScope.coroutineContext.job in outerScope.coroutineContext.job.children) {
```

Reply:
Neat! I am a little scared though, this feels brittle (why just children, and not grandchildren? Etc.)

Reply:
Yeah, I had the same thought. (Just children should be fine, thought, since we are in complete control of creating the new job, even if we're delegating to `CoroutineScope` to do so.)

I'm still stuck, though, on how to write the last unit test. By using either `CoroutineScope` builder here, I'm obscuring any visibility of the `CoroutineContext` created by the interceptor, since properties like `CoroutineName` aren't inherited. Do you have any tricks to teach me there?

Reply:
If you use `CoroutineScope(coroutineContext)`, that scope will point to the exact same context instance as the side effect gets. Not sure I understand why this wouldn't work? I guess I should read that test…

Reply:
Back to this, and I've forgotten how to replace 

```kotlin
coroutineScope { sideEffect() }
```

with 

```kotlin
val scope = CoroutineScope(coroutineContext)
```

How exactly do I use that scope to call `sideEffect`?